system-addon/lib/DownloadsManager.jsm

```diff
@@ -0,0 +1,179 @@
+ChromeUtils.import("resource://gre/modules/Services.jsm");
+Cu.importGlobalProperties(["URL"]);
+
+const {actionTypes: at} = ChromeUtils.import("resource://activity-stream/common/Actions.jsm", {});
+
+ChromeUtils.defineModuleGetter(this, "DownloadsViewUI",
+  "resource:///modules/DownloadsViewUI.jsm");
+ChromeUtils.defineModuleGetter(this, "DownloadsCommon",
+  "resource:///modules/DownloadsCommon.jsm");
+
+const RECENT_DOWNLOAD_THRESHOLD = 36 * 60 * 60 * 1000;
+
+class DownloadElement extends DownloadsViewUI.DownloadElementShell {
+  constructor(download, browser) {
+    super();
+    this._download = download;
+    this.element = browser;
+    this.element._shell = this;
+  }
+
+  get download() {
+    return this._download;
+  }
+
+  get downloadedSince() {
+    return Date.now() - this.download.endTime;
+  }
+
+  get fileType() {
+    if (!this.download.target.path) {
+      return null;
+    }
+    let items = this.download.target.path.split(".");
+    return items[items.length - 1].toUpperCase();
+  }
+
+  downloadsCmd_copyLocation() {
+    let clipboard = Cc["@mozilla.org/widget/clipboardhelper;1"].getService(Ci.nsIClipboardHelper);
+    clipboard.copyString(this.download.source.url);
+  }
+
+  downloadsCmd_openReferrer() {
+    this.element.openNewTabWith(this.download.source.referrer);
+  }
+}
+
+this.DownloadsManager = class DownloadsManager {
+  constructor(store) {
+    this._downloadData = null;
+    this._window = null;
+    this._store = null;
+    this._viewableDownloadItems = new Map();
+  }
+
+  formatDownload(element) {
+    const downloadedItem = element.download;
+    let description;
+    if (element.fileType) {
+      // If we have a file type: '1.5 MB — PNG'
+      description = `${element.sizeStrings.stateLabel} \u2014 ${element.fileType}`;
+    } else {
+      // If we do not have a file type: '1.5 MB'
+      description = `${element.sizeStrings.stateLabel}`;
+    }
+    return {
+      hostname: new URL(downloadedItem.source.url).hostname,
+      url: downloadedItem.source.url,
+      path: downloadedItem.target.path,
+      title: element.displayName,
+      description,
+      referrer: downloadedItem.source.referrer,
+      date_added: downloadedItem.endTime
+    };
+  }
+
+  init(store) {
+    this._store = store;
+    this._browser = Services.appShell.hiddenDOMWindow;
+    this._downloadData = DownloadsCommon.getData(this._browser.ownerGlobal, true, false, true);
+    this._downloadData.addView(this);
+  }
+
+  onDownloadAdded(download) {
+    const elem = new DownloadElement(download, this._browser);
+    const downloadedItem = elem.download;
+    if (!this._viewableDownloadItems.has(downloadedItem.source.url)) {
+      if (downloadedItem.succeeded && downloadedItem.target.exists) {
```

Comment:
Between these two if blocks, add `await downloadedItem.refresh();` and that should get us the correct `exists` value.

Reply:
Probably add a comment too about why